[Kleon Ang] iP

kleonang opened the pull request:
# Duke™
> "Object-oriented programming is an exceptionally bad idea which could only have originated in California." 
> – Edsger W. Dijkstra ([Source](https://www.cs.scranton.edu/~mccloske/dijkstra_quotes.html))

Duke _wipes_ your mind of having to remember things you need to do. It is 
- Text-based
- Easy to learn
- ~~FAST~~ ~~SUPER FAST~~ BLAZING FAST to use

All you need to do is:
1. Download it from [here](https://github.com/kleonang/ip/releases/download/A-Release/duke.jar)
2. Double-click it
3. Add your tasks
4. ~~Hope it doesn't crash and~~ Let it manage your tasks for you 😊

What's more, it's **free*** and _currently_ **open-source****!

Features:
- [x] Add tasks
- [x] List tasks
- [x] Delete tasks
- [x] Find tasks
- [x] Mark tasks as done
- [x] Assign priority to tasks
- [x] Interactive GUI 
- [ ] Enlightens you with the meaning of life
#
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
/**
 * Compulsory JavaDocs to avoid getting arrested by the Coding Standard™ Police.
 *
 * @param args Command-line arguments.
 */
 public static void main(String[] args) {
     new Duke("duke.txt").run();
 }
```
#
<sub>
* For the one (1) month trial period. At the end of the one (1) month trial, you will be charged the standard monthly Duke subscription price of $19.99 USD per month.
</sub>  
<br>
<sub>
** Patent pending. All rights reserved. Duke LLC has extensive legal resources at its disposal.
</sub>

Commit:
Add Gradle support

kleonang committed:
Add greet, echo and exit

kleonang committed:
Add printReply, add and list

kleonang committed:
Add Task class

kleonang committed:
Add mark as done

kleonang committed:
Change string representation of Task

kleonang committed:
Add Deadline, Event and Todo classes

kleonang committed:
Add support for tracking Todos, Deadlines and Events

kleonang committed:
Add test cases for Levels 1 to 4

kleonang committed:
Add permissions to execute runtest.sh

kleonang committed:
Add DukeException class

kleonang committed:
Add error handling

kleonang committed:
Add delete task functionality

kleonang committed:
Add JavaDocs

kleonang committed:
Change switch case indentation to match style guide

kleonang committed:
Add local saving of data

kleonang committed:
Change path of data file

kleonang committed:
Add dates and times

kleonang committed:
Add tests for datetime

kleonang committed:
Merge branch 'branch-Level-7'
Merge save feature

kleonang committed:
Merge branch 'branch-Level-8'

Merge datetime functionality

kleonang committed:
Fix datetime compatibility with saved data

kleonang committed:
Refactor code with more OOP

kleonang committed:
Add duke package

kleonang committed:
Add JUnit tests for Ui and Parser

kleonang committed:
Change list access modifier and test for ParserTest

kleonang committed:
Add and update JavaDocs

kleonang committed:
Update to adhere to Java coding standard

kleonang committed:
Add find Task functionality

kleonang committed:
Merge branch 'branch-A-CodingStandard'
Merge branch-A-CodingStandard

kleonang committed:
Merge branch-A-JavaDoc

kleonang committed:
Merge branch 'branch-Level-9'
Merge branch-Level-9

kleonang committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle
Merge add-gradle-support into branch-A-Gradle

kleonang committed:
Add gradle support

kleonang committed:
Modify annotation for gradle support

kleonang committed:
Add checkstyle rules

kleonang committed:
Modify files to conform to checkstyle rules

kleonang committed:
Merge branch 'branch-A-CheckStyle'
Merge branch-A-CheckStyle

Joel-Sung commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+    public void run() {
+        while (true) {
+            try {
+                String readIn = this.ui.nextLine();
```
I think it would be good to use nouns for variables. For example, instead of readIn, you could use something like userInput.

Joel-Sung commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,187 @@
+            }
+            default:
+                throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+            }
```
For the switch statement, I do not think it is necessary to put curly braces around each case. As it is not a standard coding convention, it may confuse readers on what the curly braces is for. However, if you think they help readability, then I think it is okay to keep it.

Joel-Sung commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,187 @@
+    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-d H:mm");
+    private final TaskList tasks;
+    private boolean toRewriteData;
+    private boolean isExit;
```
I think changing the name to 'toExit' is more intuitive as the variable is to keep track when Duke needs to exit. What do you think?

samyipsh replied:
agreed, but sticking to convention of boolean naming convention perhaps isClosed / isClosing / isExiting

kleonang replied:
Yes agreed, I'll be changing this field name to `toExit` to make more linguistic sense. I think it should also conform to the naming convention given that other similar prefixes like `is`, `has`, and `was` are also accepted. Thanks for your input everyone!

Joel-Sung commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,187 @@
+        doneTask.markAsDone();
+        Ui.printReply("Nice! I've marked this task as done:\n  " + doneTask);
+    }
+
```
I think it would be good to add JavaDocs to list and done. They do not seem like trivial methods.

Joel-Sung commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,187 @@
+    public boolean toExit() {
+        return this.isExit;
+    }
+
```
I think for methods, maybe try to keep their names as verbs, such as needsRewrite and needsToExit.

samyipsh replied:
Agreed, since its a getter could also consider more default naming convention like getExitStatus(), getRewriteStatus()

Joel-Sung commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,53 @@
+     *
+     * @return A string containing the next line of user input.
+     */
+    public String nextLine() {
```
I think it is better to name your methods as a verb to improve readability. In this case, the name could be getNextLine.

Joel-Sung reviewed (commented):
Overall, LGTM. Maybe just some method and variable names could be changed. The code is generally really well-written and follows the coding standard. Good job! 😺

kleonang committed:
Add GUI using JavaFX

kleonang committed:
Add chatbot images

samyipsh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,187 @@
+        Ui.printReply(tasksBuilder.toString());
+    }
+
+    private void done(int counter) throws DukeException {
```
"counter" varName might be a bit ambiguous, perhaps "TaskIndex" would be more intuitive

samyipsh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,187 @@
+            this.isExit = true;
+        } else if (readIn.equals("list")) {
+            list(this.tasks);
+        } else {
```
perhaps last else block is unnecessary and can be removed to neaten code logic.

samyipsh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,187 @@
+                int counter = Integer.parseInt(arguments);
+                done(counter);
+                break;
+            }
```
As joel mentioned, might be good to stick to convention and not put {} around switch, case blocks

samyipsh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,187 @@
+                String description = splitTask[0];
+                String atString = splitTask[1];
+                try {
+                    LocalDateTime at = LocalDateTime.parse(atString, FORMATTER);
```
can reconsider naming of var "at", perhaps better would be "eventTime" / "dateTime" / "time"

samyipsh commented on src/main/java/duke/Reply.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class Reply {
+    private static final String LINE = "\t____________________________________________________________";
+    private final String content;
```
final var name convention -> may be good to rename as all caps: CONTENT

samyipsh commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+package duke;
+
+import java.io.BufferedWriter;
```
nice very organised imports

samyipsh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @author Kleon Ang
+ */
+public class TaskList extends ArrayList<Task> {
```
interesting way to go bout w TaskList, perhaps a composition might be more suitable in case you want to add other details inside TaskList like timeAccessed juz to name a random example. With inheritance, its somewhat more limiting if you follow the logic that TaskList is an ArrayList (imo)

kleonang replied:
I think it works to add local fields to TaskList as an inherited ArrayList if we require additional variables, plus inheritance allows me to use the predefined iterator as well as `ArrayList::add` and `ArrayList::remove` functions for testing too 😅. Nevertheless, I'll keep your suggestion in mind in case this change is necessary in the future. Thanks!

samyipsh reviewed (commented):
Hi! nice code very neat and organized. Esp like the import, package and clarity of logic statements. Just a few tweaks and LGTM 😄

kleonang committed:
Add padding for GUI

kleonang committed:
Merge branch-Level-10

kleonang commented:
> Overall, LGTM. Maybe just some method and variable names could be changed. The code is generally really well-written and follows the coding standard. Good job! 😺

Thanks for your PR review! I'll make amendments based on your insightful comments.

kleonang commented:
> Hi! nice code very neat and organized. Esp like the import, package and clarity of logic statements. Just a few tweaks and LGTM 😄

Thanks for your PR review! I'll make amendments based on your insightful comments.

kleonang committed:
Modify code based on PR review suggestions

kleonang committed:
Rename user input string

kleonang committed:
Change mainClassName to Launcher for JAR creation

kleonang committed:
Add Java assert statements

kleonang committed:
Remove magic numbers and some hardcoded strings
There are some arguments, particularly in the Ui class, that contain
magic numbers and unintuitive hardcoded strings.

The magic numbers and some hardcoded strings do not give anyone reading
the code any context about their role in the program.

Magic numbers and the hardcoded strings that are not self-explanatory
have been removed.

Removing these magic numbers and hardcoded strings will enhance
readability of the code.

kleonang committed:
Enable finding matching tasks using Java streams
The Parser class uses a double for-loop to iterate through the tasks and
keywords to find matching strings.

This may be harder to comprehend as the two for-loops are iterating
through different data types.

Let's make the code more functional using two Java streams.

Using streams can make the code easier to understand and reduce
unintended side effects.

kleonang committed:
Merge pull request #1 from kleonang/branch-A-Assertions

Add Java assert statements

kleonang committed:
Merge branch 'master' into branch-A-CodeQuality

kleonang committed:
Merge branch 'master' into branch-A-Streams

kleonang committed:
Merge pull request #2 from kleonang/branch-A-CodeQuality

Merge branch-A-CodeQuality into master

kleonang committed:
Merge branch 'master' into branch-A-Streams

kleonang committed:
Merge pull request #3 from kleonang/branch-A-Streams

branch-A-Streams

kleonang committed:
Abstract out finding tasks into a private method

kleonang committed:
Add priority to tasks

kleonang committed:
Merge branch 'branch-C-Priority'

kleonang committed:
Modify fonts and text elements

kleonang committed:
Resize image

kleonang committed:
Improve GUI elements with FXML

kleonang committed:
Rectify checkstyle and JavaFX version issues

kleonang committed:
Add screenshot of Ui

kleonang committed:
Update screenshot of Ui

kleonang committed:
Modify header comments to comply with coding standard

kleonang committed:
Merge branch 'master' of https://github.com/kleonang/ip

kleonang committed:
Update User Guide

Add overall description and command usage guide

kleonang committed:
Update README.md

kleonang committed:
Set theme jekyll-theme-slate

kleonang committed:
Set theme jekyll-theme-midnight

kleonang committed:
Set theme jekyll-theme-minimal

kleonang committed:
Set theme jekyll-theme-tactile

kleonang committed:
Update README.md

kleonang committed:
Improve GUI UX

kleonang committed:
Merge branch 'master' of https://github.com/kleonang/ip

kleonang committed:
Update README.md

kleonang committed:
Update screenshot with refined Ui

kleonang committed:
Update Ui screenshot

kleonang committed:
Update screenshot with refined Ui

kleonang committed:
Merge branch 'master' of https://github.com/kleonang/ip

kleonang committed:
Delete Ui.png

kleonang committed:
Upload Ui screenshot

kleonang committed:
Enable case-insensitive priority levels

kleonang committed:
Fix cross-platform incompatibility

kleonang committed:
Merge branch 'master' of https://github.com/kleonang/ip

kleonang committed:
Split commands into individual classes

kleonang committed:
Abstract out methods in Storage
